chain/events/filter/event.go

```diff
@@ -0,0 +1,394 @@
+package filter
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+
+	cstore "github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+const indexed uint8 = 0x01
+
+type EventFilter struct {
+	id         string
+	minHeight  abi.ChainEpoch // minimum epoch to apply filter or -1 if no minimum
+	maxHeight  abi.ChainEpoch // maximum epoch to apply filter or -1 if no maximum
+	tipsetCid  cid.Cid
+	addresses  []address.Address   // list of actor ids that originated the event
+	keys       map[string][][]byte // map of key names to a list of alternate values that may match
+	maxResults int                 // maximum number of results to collect, 0 is unlimited
+
+	mu        sync.Mutex
+	collected []*CollectedEvent
+	lastTaken time.Time
+	ch        chan<- interface{}
+}
+
+var _ Filter = (*EventFilter)(nil)
+
+type CollectedEvent struct {
+	Event     *types.Event
+	EventIdx  int // index of the event within the list of emitted events
+	Reverted  bool
+	Height    abi.ChainEpoch
+	TipSetKey types.TipSetKey // tipset that contained the message
+	MsgIdx    int             // index of the message in the tipset
+	MsgCid    cid.Cid         // cid of message that produced event
+}
+
+func (f *EventFilter) ID() string {
+	return f.id
+}
+
+func (f *EventFilter) SetSubChannel(ch chan<- interface{}) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.ch = ch
+	f.collected = nil
+}
+
+func (f *EventFilter) ClearSubChannel() {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.ch = nil
+}
+
+func (f *EventFilter) CollectEvents(ctx context.Context, te *TipSetEvents, revert bool) error {
+	if !f.matchTipset(te) {
+		return nil
+	}
+
+	ems, err := te.messages(ctx)
+	if err != nil {
+		return xerrors.Errorf("load executed messages: %w", err)
+	}
+	for msgIdx, em := range ems {
+		for evIdx, ev := range em.Events() {
+			if !f.matchAddress(ev.Emitter) {
```

Comment:
There was an error in the spec, which Alex pointed out. The Emitter will be an ActorID, not an Address. So we need to resolve the supplied address into an ActorID in the Eth API before we feed it further onwards.